### ISSUE TITLE
Curly vector plot ?

### ISSUE THREAD

**Issue by fcodron** [NCAR/GeoCAT-examples#314]
Hello,

Is there any plan to implement "curly vector" plots in some way for "standard" Python (matplotlib, cartopy...) ?
I guess through a modification of streamlines ?

This plot type is fairly unique to NCL (as far as I know), and very useful especially for oceanography plots, where strong currents tend to flow along the coastlines, and vectors ending over land are not very nice...  

Thanks,

Francis

**Comment by michaelavs** [comment]
Hi Francis, 
I don't think there is an explicit plan in place for implementing the curly vector, but it could become part of later releases of GeoCAT software. I've been looking into if there is an already established Python version for curly vectors and it looks like currently the closest would be streamlines (which as you probably know, is slightly different than the curly vector from NCL). I'll bring this issue up with the rest of the team later this week and see what they think about it and get back to you then!

**Comment by fcodron** [comment]
Hello,

Thanks for replying !

I am only aware of this try by Kieran Hunt (U. Reading):
https://github.com/kieranmrhunt/curved-quivers
Which is I think a modification of standard MPL streamlines. I am not sure
if it can be made to work with maps. It was inspired by this stackoverflow
thread :
https://stackoverflow.com/questions/51843313/flow-visualisation-in-python-using-curved-path-following-vectors

Otherwise the curly vect seems pretty unique to NCL...

I'm attaching the figure of IPSL currents in the Arctic so you see what I
mean, perfect for ocean eddies too !

Thanks,

Francis

[image: Arctic.png]

**Comment by pilotchute** [comment]
@fcodron I've just been assigned to look into this. 
From my understanding, you curly vectors are arrows in a grid, that are scaled by the magnitude and curved based on the curl derivative of the field. 
So I think, it's the curl of the arrows that differentiates it from a standard quiver plot.
Though looking closely at the IDL output for curly vectors it actually seems more similar to broken up lines of a streamplot(), with the broken up arrows scaled by magnitude. 

Could you in your own words (or with a sketch) show what you'd be looking for in a curly vector plot?

**Comment by fcodron** [comment]
Hello,

Nice to hear you'll get into this !

Curly vectors are originally part of NCL visualization (don't know about IDL), where it's an option in vector plots. You can see an example image here : [https://www.ncl.ucar.edu/Applications/Images/vector_4_lg.png](url)

The script used to create the figure is [https://www.ncl.ucar.edu/Applications/Scripts/vector_4.ncl](url) (key argument is the @vcGlyphStyle = "CurlyVector" option).

The python "translation" here [https://geocat-examples.readthedocs.io/en/latest/gallery/Vectors/NCL_vector_4.html#sphx-glr-gallery-vectors-ncl-vector-4-py](url)
using standard vectors is much less nice...

I am not sure how exactly curly vectors were coded in NCL; they have a good documentation online but not the source code. At first glance it seems indeed similar to streamlines plot (streamplot), but with arrows at the end and a a length scaled by magnitude. I guess there would be defaults or options for density, scaling, and non-overlapping conditions.

Some attempts for implementation are described in the stackoverflow discussion above, but ideally these would work also with map projections (Cartopy...). I think these curly vectors are particularly useful when the flow is constrained by borders, like ocean currents following (but not crossing) the coast.

Francis

**Comment by pilotchute** [comment]
Thanks for the prompt reply. 
On a more conversational tone, given that python is a different software environment with a separate toolset, I'm wanting to find out what is valuable about the curly vector plot, and what sets it apart, so that I can focus on those parts of the implementation, without blindly duplicating the output of the NCL or IDL functions. 

I'm attaching a sketch of a possible way to make a curly vector plot, please let me know what you think. 
![0A4B742D-9E4F-4F88-B536-4192FAEE0416](https://user-images.githubusercontent.com/20648306/191810759-8c373a52-a0fb-40cd-9a3d-57b8a008e0d0.jpg)

**Comment by pilotchute** [comment]
the notes on the image are:
1) some number to define resolution, if different from grid (useful in high grid densities where we may not want too many lines)
2) use the grid data to create a field through interpolation (probably bi-cubic but any second order continuous interpolation should be fine)
3) for each grid point, (or resolution point) use the gradient of the field to forward and back propagate a line, for some user defined or automated (based on maximum grid value) time, which will become a distance that varies with field strength once used with the field gradient. 

 This seems to be what the [ncl plot](https://www.ncl.ucar.edu/Applications/Images/vector_4_lg.png) is doing.

**Comment by ivicajan** [comment]
My vote for curly vectors!
This was one of the reasons I started a long time ago with NCL and then pyNGL.
Think that would make GEOCAT unique and in the top of the game. Using them still with the old version of the pyNGL and shouldn't be a problem (a naive comment) to capitalize from pyNGL if there is already part of the code (or it was using NCL engine).

Cheers
IVica

**Comment by erogluorhan** [comment]
+1 curly vectors have been a much desired plotting for a long while in Python; would make a great addition -GeoCAT-examples (and maybe -viz)

**Comment by kafitzgerald** [comment]
Yeah, it does seem like there's a good bit of interest in this type of plot and not a lot in terms of Python solutions.  

Unfortunately, pyngl relies upon NGL for this and there's not an out-of-the-box solution over in matplotlib (see matplotlib/matplotlib#20038), but perhaps that's all the more reason to address it in geocat-viz?  There is an open issue over there as well in NCAR/geocat-viz#4.

**Comment by rupingmo** [comment]
I miss the curly vectors of NCL, and come frequently to GEOCAT to check its availability. Hopeful we can have it in 2026.

**Comment by areminusfarestar** [comment]
I'm just going to re-up this.  There isn't a day that goes by that I don't curse the fact that python doesn't support curly vectors.  It's one of the only reasons for me to continue to use ncl.  I'm pretty sure I could track down the original developers of that code if it would help.

**Comment by marceloandrioni** [comment]
Just wanted to leave a quick note about a package I came across while looking for something else: [Skyborn](https://github.com/QianyeSu/Skyborn?tab=readme-ov-file#curly-vector-plotting). According to the author, the curly vector plotting functionality is inspired by NCL.

![first](https://raw.githubusercontent.com/QianyeSu/Skyborn/main/docs/source/images/wrf_curly_vector_integration_direction_density08_ref10.png)

![second](https://raw.githubusercontent.com/QianyeSu/Skyborn/main/docs/source/images/hadley_msf_v_omega_ncl_curly.png)

**Comment by areminusfarestar** [comment]
you are a hero.
-m

**Comment by kafitzgerald** [comment]
I believe Ultraplot has also made some progress on this: https://github.com/Ultraplot/UltraPlot/discussions/370
